components/batch-system/src/channel.rs

```diff
@@ -0,0 +1,252 @@
+{
+    type Fsm = C;
+
+    fn consume_msg_resource(&self, _msg: &<Self::Fsm as Fsm>::Message) {}
+
+    #[inline]
+    fn schedule(&self, fsm: Box<C>) {
+        match self.sender.send(FsmTypes::Control(fsm)) {
+            Ok(()) => {}
+            Err(SendError(FsmTypes::Control(fsm))) => warn!("failed to schedule fsm {:p}", fsm),
+            _ => unreachable!(),
+        }
+    }
+
+    fn shutdown(&self) {
+        // TODO: close it explicitly once it's supported.
+        // Magic number, actually any number greater than poll pool size works.
+        for _ in 0..256 {
+            let _ = self.sender.send(FsmTypes::Empty);
+        }
+    }
+}
+
+pub enum FsmSender<N: Fsm, C: Fsm> {
+    Vanilla(channel::Sender<FsmTypes<N, C>>),
+    Priority {
+        resource_ctl: Arc<ResourceController>,
+        sender: priority_queue::Sender<FsmTypes<N, C>>,
+        last_msg_group: RefCell<String>,
+    },
+}
+
+impl<N, C> Clone for FsmSender<N, C>
+where
+    N: Fsm,
+    C: Fsm,
+{
+    fn clone(&self) -> Self {
+        match self {
+            FsmSender::Vanilla(sender) => FsmSender::Vanilla(sender.clone()),
+            FsmSender::Priority {
+                resource_ctl,
+                sender,
+                ..
+            } => FsmSender::Priority {
+                resource_ctl: resource_ctl.clone(),
+                sender: sender.clone(),
+                last_msg_group: RefCell::new(String::new()),
+            },
+        }
+    }
+}
+
+impl<N: Fsm, C: Fsm> FsmSender<N, C> {
+    pub fn send(&self, fsm: FsmTypes<N, C>) -> Result<(), SendError<FsmTypes<N, C>>> {
+        match self {
+            FsmSender::Vanilla(sender) => sender.send(fsm),
+            FsmSender::Priority {
+                resource_ctl,
+                sender,
+                last_msg_group,
+            } => {
+                // TODO: pass different priority
+                let pri = resource_ctl
+                    .get_priority(last_msg_group.borrow().as_bytes(), CommandPri::Normal);
+                sender.send(fsm, pri)
+            }
+        }
+    }
+
+    pub fn try_send(&self, fsm: FsmTypes<N, C>) -> Result<(), TrySendError<FsmTypes<N, C>>> {
+        match self {
+            FsmSender::Vanilla(sender) => sender.try_send(fsm),
+            FsmSender::Priority {
+                resource_ctl,
+                sender,
+                last_msg_group,
+            } => {
+                let priority = resource_ctl
+                    .get_priority(last_msg_group.borrow().as_bytes(), CommandPri::Normal);
+                sender.try_send(fsm, priority)
+            }
+        }
+    }
+
+    fn consume_msg_resource(&self, msg: &N::Message) {
+        match self {
+            FsmSender::Vanilla(_) => {}
+            FsmSender::Priority {
+                resource_ctl,
+                last_msg_group,
+                ..
+            } => {
+                let mut dominant_group = "".to_owned();
+                let mut max_write_bytes = 0;
```

Comment:
Seems that this has not been put into practice, is this intended? Maybe remove it, and add it back when it is needed to avoid confusion?